### ISSUE TITLE
New rule: forbid unsafe enum assignment

### ISSUE THREAD

**Issue by danielnixon** [danielnixon/eslint-plugin-total-functions#724]
```typescript
enum ZeroOrOne {
  Zero = 0,
  One = 1,
}

// Compiles
const zeroOrOne: ZeroOrOne = 2;

const foo = (a: ZeroOrOne): void => {
  return;
};

// Compiles
foo(2);
```

**Comment by danielnixon** [comment]
You can ban enum _declaration_ with the `no-restricted-syntax` rule, like so:

```
    // Ban enums.
    // TODO make this a real rule instead of using `no-restricted-syntax`: https://github.com/danielnixon/eslint-config-typed-fp/issues/284
    "no-restricted-syntax": [
      "error",
      {
        selector: "TSEnumDeclaration",
        message: "Don't declare enums",
      },
    ],
```

Indeed, we do exactly that over in https://github.com/danielnixon/eslint-config-typed-fp/blob/master/src/index.ts#L47

But that won't help you if the enum is declared by a library and you're, for example, passing an argument to a function that expects such an enum. So we want a rule to detect the unsafe assignment whether we have banned enums or not.

**Comment by danielnixon** [comment]
Here is a POC: https://github.com/danielnixon/eslint-plugin-total-functions/pull/new/ban-enum-assignment

**Comment by danielnixon** [comment]
It might be nice to:
* not flag string enums, only numeric
* not flag assignment of number literals to numeric enum values that we can verify at compile time are safe (i.e. actually belong to the enum)

**Comment by danielnixon** [comment]
Counterargument to the second bulletpoint in above comment would be:

```ts
// Let's start with an implicit numeric enum:
enum AOrB {
  A, // Implicitly 0
  B, // Implicitly 1
}

// And then let's use it
const b: AOrB = 1; // At the time I wrote this, I believed 1 === AOrB.B

// But if someone later inserts a new value in the middle...
enum AOrB {
  A, // Implicitly 0
  C, // Implicitly 1
  B, // Implicitly 2 now!
}

// This existing code still compiles, but is now misleading/wrong, because 1 actually corresponds to AOrB.C
const b: AOrB = 1;

// If we had instead written this, there would be no issue:
const b: AOrB = AOrB.B;
```

**Comment by danielnixon** [comment]
```ts
// If we had instead written this, there would be no issue:
const b: AOrB = AOrB.B;
```

Unless of course we had a bunch of enum values stored in a database using their numeric values, in which case by adding AOrB.C we have accidentally changed their meaning in a disastrous way. Another argument against enums _at all_.
